[CXF-8361] Finalized support for Jakarta NS

ropalka opened the pull request:
https://issues.apache.org/jira/browse/CXF-8361
 * introduced new org.apache.cxf.jaxws.handler.jakarta_types package for JAXB de/serialization
 * added new methods in HandlerChainBuilder
 * enhanced handlers de/serialization tests

reta commented:
Thanks for the PR, @ropalka, I think we got to the point when it is probably make sense to completely separate Jakartaee and Javaee name handler chains (I think you would agree). I suggest to:
1. Introduce `org.apache.cxf.jaxws.handler.jakartaee` and `org.apache.cxf.jaxws.handler.javaee`
2. Introduce `JakartaeeHandlerChainBuilder` and `JavaeeHandlerChainBuilder`
3. Basically, I think it safe to assume that either JakartaEE or JavaEE namespace is going to be used, so in this case `AnnotationHandlerChainBuilder` could inspect the namespace, instantiate  `JakartaeeHandlerChainBuilder` or`JavaeeHandlerChainBuilder`, and delegate to one of those. In this case, the conditionals should be gone and implementation becomes much cleaner. 

What do you think?

ropalka commented:
Hi @reta . It makes sense to me. What I am worry about is backward compatibility.
Isn't jaxws frontend considered to be a public API? In other words if I would refactor package
from org.apache.cxf.jaxws.handler.types to org.apache.cxf.jaxws.handler.javaee
I would break compatibility.

reta commented:
Thanks for the reply @ropalka. We won't be changing the public API, which is outlined by `AnnotationHandlerChainBuilder`. The `JakartaeeHandlerChainBuilder`  and `JavaeeHandlerChainBuilder` would encapsulate the implementation specifics to each namespace, and are not supposed to be used outside `AnnotationHandlerChainBuilder`. All the existing classes, under `org.apache.cxf.jaxws.handler.types` fe, could stay were they are. What do you think?

ropalka commented:
Hi @reta . The problem with your proposal is AnnotationHandlerChainBuilder is extending HandlerChainBuilder class.
If I would introduce both JakartaeeHandlerChainBuilder and JavaeeHandlerChain builders I would need to drop
"AnnotationHandlerChainBuilder extends HandlerChainBuilder" relationship which is again not public API backward compatible change. The proposal PR I sent is 100% backward compatible.

ropalka commented:
I enhanced this PR to be as much backward compatible as possible @reta.
I introduced new _org.apache.cxf.jaxws.handler.jakartaee_ package as you proposed.
I left _org.apache.cxf.jaxws.handler.types_ package untouched because of BC.
I introduced JAXB deserialization adaptors to be able to eliminate HandlerChainBuilder
'jakartaee' methods I indroduced in previous pull request proposal. WDYT about it now?

reta commented:
> 
> 
> I enhanced this PR to be as much backward compatible as possible @reta.
> I introduced new _org.apache.cxf.jaxws.handler.jakartaee_ package as you proposed.
> I left _org.apache.cxf.jaxws.handler.types_ package untouched because of BC.
> I introduced JAXB deserialization adaptors to be able to eliminate HandlerChainBuilder
> 'jakartaee' methods I indroduced in previous pull request proposal. WDYT about it now?

@ropalka thank you for the effort, it definitely looks better but the main problem is still unsolved: the processing of the JavaEE and JakartaEE namespaces is tangled. May I ask you please to take at look at https://github.com/apache/cxf/pull/756, it is based of `master` (without your changes) but illustrates the idea of how the processing could be separated. AFAIK it does not change or alter any public APIs, only the implementation details have been moved around. Would appreciate to hear any concerns, thank you.

jimma commented:
It's good to not change AnnotationHandlerChainBuilder's public api and hierarchy.   +1 for this change.

ropalka commented:
I updated this PR as you suggested @reta, please review.

reta commented on rt/frontend/jaxws/src/main/java/org/apache/cxf/jaxws/handler/AnnotationHandlerChainBuilder.java:
```diff
@@ -177,149 +129,8 @@ public ClassLoader run() {
-            context = JAXBContext.newInstance(PortComponentHandlerType.class);
-        }
-        return context;
+    List<Handler> buildHandlerChain(PortComponentHandlerType ht) {
```
I would prefer to keep this implementation details (either `private` or even remove the function altogether), but I am perfrectly fine to make the change after merge.

ropalka replied:
It cannot be private as it is used by both JakartaeeHandlerChainBuilder & JavaeeHandlerChainBuilder.

reta commented on rt/frontend/jaxws/src/main/java/org/apache/cxf/jaxws/handler/BaseHandlerChainBuilder.java:
```diff
@@ -0,0 +1,221 @@
+abstract class BaseHandlerChainBuilder {
+    private static final String HANDLER_CHAINS_E = "handler-chains";
+    private static final String HANDLER_CHAIN_E = "handler-chain";
+    final AnnotationHandlerChainBuilder delegate;
```
Minor, I would make remove it from `BaseHandlerChainBuilder`, it is needed only in concrete implementation, will take care of it after merge if you prefer.

ropalka replied:
I have no objections @reta - cleaning it up after merge sounds good to me.

reta commented:
@ropalka thank you, LGTM, I have a few minor comments which I could address myself if you prefer, otherwise - I think we are good to go

reta commented:
Thanks again for your work, @ropalka

lulseged commented:
I have tried 3.5.0-SNAPSHOT and miss a lot of jakarta.* packages. When do you plan to move to jakarta namespaces?
The change will not be backward compatible. What is the use of making it backward compatible?
I would create 4.0.0-SNAPSHOT and start jakarta namespace from there and keep 3.x.x javax namespace.

reta commented:
@lulseged please follow https://github.com/apache/cxf/pull/737
